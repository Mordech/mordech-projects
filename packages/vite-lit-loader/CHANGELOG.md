# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.31.2](https://github.com/Mordech/mordech-projects/compare/v0.31.1...v0.31.2) (2023-12-09)

### Bug Fixes

- **vite-lit-loader:** add dependencies ([#105](https://github.com/Mordech/mordech-projects/issues/105)) ([c9d4bb7](https://github.com/Mordech/mordech-projects/commit/c9d4bb757396ba3850492d92a833fb09ab5144f8))

# [0.31.0](https://github.com/Mordech/mordech-projects/compare/v0.29.1...v0.31.0) (2023-12-09)

### Features

- **vite-lit-loader:** add attributes with query parameters ([#104](https://github.com/Mordech/mordech-projects/issues/104)) ([960e6a0](https://github.com/Mordech/mordech-projects/commit/960e6a01895c58dd45208fcb3fa7bb2d159bea3f))

# [0.30.0](https://github.com/Mordech/mordech-projects/compare/v0.29.1...v0.30.0) (2023-12-02)

**Note:** Version bump only for package @mordech/vite-lit-loader

## [0.29.1](https://github.com/Mordech/mordech-projects/compare/v0.29.0...v0.29.1) (2023-09-17)

**Note:** Version bump only for package @mordech/vite-lit-loader

# [0.29.0](https://github.com/Mordech/mordech-projects/compare/v0.28.0...v0.29.0) (2023-09-16)

### Features

- vite-lit-loader ([#97](https://github.com/Mordech/mordech-projects/issues/97)) ([df75c81](https://github.com/Mordech/mordech-projects/commit/df75c814ba237829727a43645e2d097cd0d614c9))
