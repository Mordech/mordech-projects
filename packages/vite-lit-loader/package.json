{
  "name": "@mordech/vite-lit-loader",
  "version": "0.33.3",
  "type": "commonjs",
  "main": "dist/index.js",
  "license": "MIT",
  "types": "dist/index.d.ts",
  "files": [
    "dist",
    "types"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Mordech/mordech-projects.git",
    "directory": "/packages/vite-lit-loader"
  },
  "homepage": "https://github.com/Mordech/mordech-projects/tree/main/packages/vite-lit-loader#readme",
  "dependencies": {
    "cheerio": "1.0.0-rc.12",
    "tslib": "^2.6.2"
  },
  "publishConfig": {
    "access": "public"
  }
}
