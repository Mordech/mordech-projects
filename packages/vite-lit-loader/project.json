{
  "name": "@mordech/vite-lit-loader",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "packages/vite-lit-loader/src",
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "nx:run-commands",
      "options": {
        "commands": [
          {
            "command": "tsc -p packages/vite-lit-loader/tsconfig.lib.json"
          }
        ],
        "cwd": ".",
        "parallel": false
      }
    },
    "lint": {
      "command": "eslint packages/vite-lit-loader/src"
    },
    "prepare": {
      "executor": "nx:run-commands",
      "options": {
        "commands": [
          {
            "command": "nx build @mordech/vite-lit-loader"
          }
        ],
        "cwd": "packages/vite-lit-loader",
        "parallel": false
      }
    },
    "test": {
      "executor": "@nx/vite:test",
      "outputs": ["{options.reportsDirectory}"],
      "options": {
        "reportsDirectory": "/packages/vite-lit-loader/coverage"
      }
    }
  },
  "tags": []
}
