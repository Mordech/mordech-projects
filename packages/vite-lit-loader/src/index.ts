export {
  litStyleLoader,
  litTemplateLoader,
  viteLitLoader,
} from './lib/vite-lit-loader';
