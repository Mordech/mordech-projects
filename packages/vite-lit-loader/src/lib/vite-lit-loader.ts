import * as cheerio from 'cheerio';
import { readFileSync } from 'fs';
import { Plugin } from 'vite';

/**
 * Handles loading CSS SCSS and SASS files as lit-html styles
 */
export function litStyleLoader(): Plugin {
  return {
    name: 'vite-styles-loader',
    enforce: 'post',

    transform(_, id) {
      const split = id.split('?');
      const queryParams = new URLSearchParams(split[1]);

      if (queryParams.get('lit') === null) return;

      const path = split[0];
      const fileType = path.split('.').pop();

      if (!fileType?.match(/scss|sass|css/)) return;

      return `
          import { unsafeCSS } from 'lit';
          import styles from '${path}?inline';

          export default unsafeCSS(styles);
          `;
    },
  };
}

/**
 * Handles loading HTML and SVG files as lit-html templates
 */
export function litTemplateLoader(): Plugin {
  return {
    name: 'vite-lit-loader',
    enforce: 'pre',

    async load(id) {
      const split = id.split('?');
      const queryParams = new URLSearchParams(split[1]);

      if (queryParams.get('lit') === null) return;

      const path = split[0];
      const fileType = path.split('.').pop();

      if (!fileType?.match(/html|svg/)) return;
      const file = readFileSync(path, 'utf-8');

      if (fileType === 'svg') {
        const svgElement = cheerio.load(file, { xmlMode: true });

        if (queryParams.get('as-use') !== null) {
          const svgId = svgElement('svg').attr('id');

          svgElement('svg')
            .removeAttr('id')
            .removeAttr('viewBox')
            .html(`<use href="#${svgId}"></use>`);

          addAttributesFromParams(queryParams, svgElement('svg'));

          const svgOutput = svgElement.html() || '';

          return `
          import { unsafeSVG } from 'lit/directives/unsafe-svg.js';
          
          export default unsafeSVG(\`${svgOutput}\`)
          `;
        }

        addAttributesFromParams(queryParams, svgElement('svg'));

        const svgOutput = svgElement.html() || '';

        return `
          import { unsafeSVG } from 'lit/directives/unsafe-svg.js';

          export default unsafeSVG(\`${svgOutput}\`)
          `;
      }

      if (fileType === 'html') {
        const htmlElement = cheerio.load(file, { xmlMode: true });

        addAttributesFromParams(queryParams, htmlElement('html'));

        const htmlOutput = htmlElement.html() || '';

        return `
          import { unsafeHTML } from 'lit/directives/unsafe-html.js';

          export default unsafeHTML(\`${htmlOutput}\`);
          `;
      }

      return;
    },
  };
}

/**
 * @deprecated
 * This function is deprecated.
 * Use `litStyleLoader()` and `litTemplateLoader()` instead
 */
export function viteLitLoader(): Plugin {
  return {
    name: 'vite-lit-loader',
    enforce: 'post',

    async transform(_, id) {
      if (id.endsWith('lit')) {
        const path = id.split('?')[0];
        const fileType = path.split('.').pop();

        if (!fileType?.match(/html|svg|scss|sass|css/)) return;

        if (fileType === 'svg') {
          return `
          import { unsafeSVG } from 'lit/directives/unsafe-svg.js';
          import svg from '${path}?raw';
          export default unsafeSVG(svg)
          `;
        }

        if (fileType === 'html') {
          return `
          import { unsafeHTML } from 'lit/directives/unsafe-html.js';
          import html from '${path}?raw';
          export default unsafeHTML(html)
          `;
        }

        if (fileType.match(/scss|sass|css/)) {
          const path = id.split('?')[0];

          return `
          import { unsafeCSS } from 'lit';
          import styles from '${path}?inline';
          export default unsafeCSS(styles);
          `;
        }

        return;
      }
      return;
    },
  };
}

function addAttributesFromParams(
  querystring: URLSearchParams,
  svgElement: cheerio.Cheerio<cheerio.Element>
) {
  for (const [key, value] of querystring) {
    if (key === 'lit') continue;

    if (key === 'as-use') continue;

    svgElement.attr(key, value);
  }
}
