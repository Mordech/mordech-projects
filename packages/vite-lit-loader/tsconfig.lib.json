{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "outDir": "./dist/",
    "rootDir": "./src/",
    "declaration": true,
    "types": ["node", "./types"]
  },
  "include": ["src/**/*.ts", "src/lib/vite-lit-loader.js"],
  "exclude": ["jest.config.ts", "src/**/*.spec.ts", "src/**/*.test.ts"]
}
