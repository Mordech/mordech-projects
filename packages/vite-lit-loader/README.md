# @mordech/vite-lit-loader

A vite plugin for loading `svg`, `html`, `css`, `scss`, and `sass` files as `lit-html` templates/styles.

## Install

```bash
pnpm add --dev @mordech/vite-lit-loader
yarn add --dev @mordech/vite-lit-loader
npm install --dev @mordech/vite-lit-loader
```

```ts
// vite.config.ts
import { defineConfig } from 'vite';
import { litStyleLoader, litTemplateLoader } from '@mordech/vite-lit-loader';

export default defineConfig({
  plugins: [litStyleLoader(), litTemplateLoader()],
});
```

`litStyleLoader()` Used to load CSS, SCSS, and SASS files as lit-html styles.

`litTemplateLoader()` Used to load SVG and HTML files as lit-html templates.

They are optional and can be used separately.

## Typescript

To load module types add the following to your `vite-env.d.ts`:

```ts
/// <reference types="@mordech/vite-lit-loader/types" />
```

## Usage

```ts
import svg from './icon.svg?lit';
import html from './template.html?lit';
import css from './styles.css?lit';
...
import { LitElement, html, css } from 'lit';
import { customElement, property } from 'lit/decorators.js'

@customElement('readme-component')
export class ReadmeComponent extends LitElement {
  static styles = [css];

  render() {
    return html`
      <div class="readme-component">
        <h1>Readme Component</h1>
        ${html}
        ${svg}
      </div>
    `;
  }
}
```

### Query Parameters

⚠️ When using query parameters make sure you include `&lit` at the end of the query string.

You can pass query parameters to SVG or HTML. They will be embedded into the element as attributes.

#### `?use&lit`

Creates a `<use>` element that refers to the `id` of the imported SVG. This is useful for reusing icons.

Make sure to include an `id` attribute on the original SVG element.

```ts
import svg from './icon.svg?as-use&lit';

render() {
  return html`
    <div class="readme-component">
      <h1>Readme Component</h1>
      ${svg}
      <!--
        Result:
        <svg xmlns="http://www.w3.org/2000/svg">
          <use href="#icon"></use>
        </svg>
       -->
    </div>
  `;
}
```
