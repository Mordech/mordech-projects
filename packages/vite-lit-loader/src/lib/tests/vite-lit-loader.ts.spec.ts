import { join } from 'path';

import { loadStyle, loadTemplate, transformStyle } from '../vite-lit-loader';

const path = (...args: string[]) =>
  join(
    process.cwd(),
    '/packages/vite-lit-loader/src/lib/tests/assets/',
    ...args,
  );

describe('vite-lit-loader', () => {
  describe('loadTemplate', () => {
    it('should load HTML template', async () => {
      const id = path('test.html?lit');
      const result = await loadTemplate(id);
      expect(result).toBeDefined();
      expect(result).toContain('unsafeHTML');
      expect(result).toContain('<title>Test HTML</title>');
      expect(result).toMatchSnapshot();
    });

    it('should load SVG template', async () => {
      const id = path('test.svg?lit');
      const result = await loadTemplate(id);
      expect(result).toBeDefined();
      expect(result).toContain('unsafeSVG');
      expect(result).toMatchSnapshot();
    });

    it('should not load template for unsupported file types', async () => {
      const id = path('test.png');
      const result = await loadTemplate(id);
      expect(result).toBeUndefined();
    });

    it('should not load template for unsupported query', async () => {
      const id = path('test.svg?inline');
      const result = await loadTemplate(id);
      expect(result).toBeUndefined();
    });

    it('should load SVG template as use', async () => {
      const id = path('test.svg?as-use&lit');
      const result = await loadTemplate(id);
      expect(result).toBeDefined();
      expect(result).toContain('<use href="#test-svg"');
    });

    it('should pass attributes from query', async () => {
      const id = path('test.svg?as-use&class=foo&fill=red&lit');
      const result = await loadTemplate(id);
      expect(result).toBeDefined();
      expect(result).toContain('<use href="#test-svg"');
      expect(result).toContain('class="foo"');
      expect(result).toContain('fill="red"');
    });
  });

  describe('loadStyle', () => {
    it('should load SCSS style as empty string', async () => {
      const id = 'test.scss?lit';
      const result = await loadStyle(id);
      expect(result).toBeDefined();
      expect(result).toBe('');
    });

    it('should not load style for unsupported file types', async () => {
      const id = 'test.txt?lit';
      const result = await loadStyle(id);
      expect(result).toBeUndefined();
    });

    it('should not load style for unsupported query', async () => {
      const id = 'test.scss?inline';
      const result = await loadStyle(id);
      expect(result).toBeUndefined();
    });
  });

  describe('transformStyle', () => {
    it('should transform SCSS style', () => {
      const code = 'some SCSS code';
      const id = 'test.scss?lit';
      const result = transformStyle(code, id);
      expect(result).toBeDefined();
      expect(result).toContain('unsafeCSS');
    });

    it('should not transform style for unsupported file types', () => {
      const code = 'some CSS code';
      const id = 'test.txt';
      const result = transformStyle(code, id);
      expect(result).toBeUndefined();
    });
  });
});
