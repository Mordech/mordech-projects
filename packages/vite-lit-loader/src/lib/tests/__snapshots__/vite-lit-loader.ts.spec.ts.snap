// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`vite-lit-loader > loadTemplate > should load HTML template 1`] = `
"
      import { unsafeHTML } from 'lit/directives/unsafe-html.js';

      export default unsafeHTML(\`<!doctype html>
<html>
  <head>
    <title>Test HTML</title>
  </head>
  <body>
    <h1>Hello, World!</h1>
    <p>This is a test HTML document.</p>
  </body>
</html>
\`);
      "
`;

exports[`vite-lit-loader > loadTemplate > should load SVG template 1`] = `
"
      import { unsafeSVG } from 'lit/directives/unsafe-svg.js';

      export default unsafeSVG(\`<svg xmlns="http://www.w3.org/2000/svg" id="test-svg" width="100" height="100">
  <circle cx="50" cy="50" r="40" fill="blue"/>
</svg>\`)
      "
`;
